File: MIDTERM-LAB-3/MP11_ColumnFrequency.js
/**
 * MP11 – Column Frequency Counter
 * Author: Gonzales, Chris Allen S.
 * Course: Programming 2 – Machine Problem Set
 *
 * Description:
 *   Reads the Pearson VUE exam results CSV file and counts how many times
 *   each unique value appears in a user-selected column. Useful for seeing
 *   distributions like how many took each exam, or how many passed vs. failed.
 *
 * Program Logic:
 *   1. Ask the user for the CSV file path via readline
 *   2. Read the file using Node.js fs module
 *   3. Skip metadata rows until the header row ("Candidate") is found
 *   4. Parse the header row to get column names
 *   5. Display column names and ask the user to choose one
 *   6. Count occurrences of each unique value in the chosen column
 *   7. Display a sorted frequency table with counts and percentages
 *
 * Run with: node MP11_ColumnFrequency.js
 */

// ---- Node.js built-in modules ----
const fs       = require('fs');       // File system: for reading the CSV file
const readline = require('readline'); // For terminal input/output prompts

// Create readline interface connected to terminal
const rl = readline.createInterface({
    input:  process.stdin,
    output: process.stdout
});

// ---------------------------------------------------------------
// Step 1: Ask for file path
// ---------------------------------------------------------------
rl.question('Enter the CSV dataset file path: ', (filePath) => {

    // ---------------------------------------------------------------
    // Step 2: Read the file
    // ---------------------------------------------------------------
    let fileContent;

    try {
        // Read entire file as a UTF-8 string
        fileContent = fs.readFileSync(filePath.trim(), 'utf8');
    } catch (err) {
        console.log('Error: File not found or could not be read – ' + filePath);
        rl.close();
        return;
    }

    // Split file into lines; handle both Windows and Unix line endings
    const lines = fileContent.split(/\r?\n/);

    // ---------------------------------------------------------------
    // Step 3: Parse file into headers and data rows
    // ---------------------------------------------------------------
    let headers    = null;   // Column header names from the header row
    const records  = [];     // All valid data rows
    let headerFound = false; // Flag: true once we've passed the header row

    for (const line of lines) {

        // Skip metadata lines until the header row is found
        // The header row is identified by containing the word "Candidate"
        if (!headerFound) {
            if (line.includes('Candidate')) {
                headerFound = true;
                // Parse the header row to extract column names
                headers = parseCSVLine(line);
            }
            continue;  // Skip this line (it's metadata or the header itself)
        }

        // Skip blank lines
        if (line.trim() === '') continue;

        // Parse data row
        const fields = parseCSVLine(line);

        // Must have at least 8 columns to be valid
        if (fields.length < 8) continue;

        // Skip rows with empty candidate name (footer rows)
        if (fields[0].trim() === '') continue;

        records.push(fields);  // Store the valid row
    }

    // Guard: no data or headers found
    if (!headers || records.length === 0) {
        console.log('Error: No valid data found in the file.');
        rl.close();
        return;
    }

    // ---------------------------------------------------------------
    // Step 4: Display column list and ask user to choose one
    // ---------------------------------------------------------------
    console.log('\nAvailable Columns:');
    console.log('──────────────────────────────');

    // Show only columns that have a non-empty header name
    headers.forEach((col, index) => {
        if (col.trim() !== '') {
            console.log(`  [${index}] ${col.trim()}`);
        }
    });

    console.log('──────────────────────────────');

    rl.question('Enter column number to analyze: ', (input) => {
        rl.close();  // Close input after receiving the column choice

        // Parse the user's column selection
        const selectedCol = parseInt(input.trim(), 10);

        // Validate the selected column index
        if (isNaN(selectedCol) || selectedCol < 0 || selectedCol >= headers.length) {
            console.log('Error: Invalid column number.');
            return;
        }

        const selectedColumnName = headers[selectedCol].trim();

        // ---------------------------------------------------------------
        // Step 5: Count frequency of each unique value in the chosen column
        // ---------------------------------------------------------------

        // Object used as a frequency map: { value: count }
        const frequencyMap = {};

        for (const row of records) {
            // Get cell value; use "(empty)" if column doesn't exist or is blank
            const cellValue = (selectedCol < row.length && row[selectedCol].trim() !== '')
                              ? row[selectedCol].trim()
                              : '(empty)';

            // Increment count for this value (initialize to 0 if not yet seen)
            frequencyMap[cellValue] = (frequencyMap[cellValue] || 0) + 1;
        }

        // Convert map to array and sort by count descending
        const sortedEntries = Object.entries(frequencyMap)
            .sort((a, b) => b[1] - a[1]);  // Sort by count, highest first

        const totalRows = records.length;  // Used to calculate percentages

        // ---------------------------------------------------------------
        // Step 6: Display the frequency table
        // ---------------------------------------------------------------
        console.log();
        console.log('╔══════════════════════════════════════════════════════════╗');
        console.log('║           MP11 – COLUMN FREQUENCY COUNTER                ║');
        console.log('╚══════════════════════════════════════════════════════════╝');
        console.log('  Source   : Pearson VUE – UPHSD Molino Exam Results');
        console.log(`  Column   : ${selectedColumnName}`);
        console.log(`  Total Rows Analyzed: ${totalRows}`);
        console.log('──────────────────────────────────────────────────────────');

        // Table header row
        const valueHeader   = 'Value'.padEnd(54);
        const countHeader   = 'Count'.padStart(6);
        const percentHeader = 'Percent'.padStart(8);
        console.log(`  ${valueHeader}  ${countHeader}  ${percentHeader}`);
        console.log('  ' + '─'.repeat(72));

        for (const [value, count] of sortedEntries) {
            const percent  = ((count / totalRows) * 100).toFixed(2);  // Percentage to 2 decimal places
            const padValue = value.padEnd(54);          // Left-align value
            const padCount = String(count).padStart(6); // Right-align count
            const padPct   = `${percent}%`.padStart(8); // Right-align percent
            console.log(`  ${padValue}  ${padCount}  ${padPct}`);
        }

        console.log('  ' + '─'.repeat(72));

        // Totals row
        const totalLabel = 'TOTAL'.padEnd(54);
        const totalCount = String(totalRows).padStart(6);
        const totalPct   = '100.00%'.padStart(8);
        console.log(`  ${totalLabel}  ${totalCount}  ${totalPct}`);

        console.log('══════════════════════════════════════════════════════════');
        console.log('           End of Column Frequency Report                 ');
        console.log('══════════════════════════════════════════════════════════');
    });
});

// ---------------------------------------------------------------
// Helper function
// ---------------------------------------------------------------

/**
 * parseCSVLine – Parses a single CSV line into an array of string fields.
 *
 * Handles fields enclosed in double-quotes such as "LastName,FirstName".
 * The comma inside the quotes is treated as part of the field, not a separator.
 *
 * @param {string} line  - A raw CSV line
 * @returns {string[]}   - Array of string field values
 */
function parseCSVLine(line) {
    const fields = [];    // Stores each parsed field
    let current  = '';    // Builds up the current field
    let inQuotes = false; // True when parsing inside double-quotes

    for (const char of line) {
        if (char === '"') {
            inQuotes = !inQuotes;  // Toggle in/out of quoted mode
        } else if (char === ',' && !inQuotes) {
            fields.push(current);  // Comma outside quotes = field separator
            current = '';          // Reset for the next field
        } else {
            current += char;       // Regular character: append to current field
        }
    }

    fields.push(current);  // Save the final field (no trailing comma)
    return fields;
}

File: MIDTERM-LAB 3/MP11_ColumnFrequency.java
import java.io.*;
import java.util.*;

/**
 * MP11 – Column Frequency Counter
 * Author: Gonzales, Chris Allen S.
 * Course: Programming 2 – Machine Problem Set
 *
 * Description:
 *   Reads the Pearson VUE exam results CSV file and counts how many times
 *   each unique value appears in a user-selected column. The user can choose
 *   any column (e.g., Exam, Result, Type) to see a frequency distribution.
 *
 * Program Logic:
 *   1. Ask the user for the CSV file path
 *   2. Skip metadata rows until the real header row ("Candidate") is found
 *   3. Display the available column names and ask the user to pick one
 *   4. Parse each data row and collect values from the selected column
 *   5. Count occurrences of each unique value using a Map
 *   6. Display a sorted frequency table with counts and percentages
 */
public class MP11_ColumnFrequency {

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        // Prompt user to enter the path to the CSV file
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();

        // ---------------------------------------------------------------
        // Step 1: Load all data rows into memory
        // ---------------------------------------------------------------
        String[] headers = null;          // Array of column header names
        List<String[]> records = new ArrayList<>();  // All valid data rows

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line;
            boolean headerFound = false;  // Flag: have we passed the header row?

            while ((line = br.readLine()) != null) {

                // Skip metadata lines until the real header row is found
                // Header row contains the word "Candidate"
                if (!headerFound) {
                    if (line.contains("Candidate")) {
                        headerFound = true;
                        // This line IS the header row; parse it for column names
                        headers = parseCSVLine(line);
                    }
                    continue;  // Don't treat this line as a data row
                }

                // Skip blank lines
                if (line.trim().isEmpty()) continue;

                // Parse the data row
                String[] fields = parseCSVLine(line);

                // Must have at least 8 columns to be valid
                if (fields.length < 8) continue;

                // Skip rows with empty candidate name (footer rows)
                if (fields[0].trim().isEmpty()) continue;

                records.add(fields);  // Store the valid row
            }

        } catch (FileNotFoundException e) {
            System.out.println("Error: File not found – " + filePath);
            return;
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return;
        }

        // Guard: file was empty or no headers found
        if (headers == null || records.isEmpty()) {
            System.out.println("Error: No valid data found in the file.");
            return;
        }

        // ---------------------------------------------------------------
        // Step 2: Display columns and let user choose one
        // ---------------------------------------------------------------
        System.out.println();
        System.out.println("Available Columns:");
        System.out.println("──────────────────────────────");

        for (int i = 0; i < headers.length; i++) {
            // Only show columns that have a non-empty header name
            if (!headers[i].trim().isEmpty()) {
                System.out.printf("  [%d] %s%n", i, headers[i].trim());
            }
        }

        System.out.println("──────────────────────────────");
        System.out.print("Enter column number to analyze: ");

        int selectedCol;  // The column index chosen by the user

        try {
            selectedCol = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            System.out.println("Error: Please enter a valid column number.");
            return;
        }

        // Validate the column number is within range
        if (selectedCol < 0 || selectedCol >= headers.length) {
            System.out.println("Error: Column number out of range.");
            return;
        }

        String selectedColumnName = headers[selectedCol].trim();

        // ---------------------------------------------------------------
        // Step 3: Count the frequency of each unique value in the column
        // ---------------------------------------------------------------

        // Map: unique value -> how many times it appears in the selected column
        Map<String, Integer> frequencyMap = new LinkedHashMap<>();

        for (String[] row : records) {
            // Safely get the cell value; default to "(empty)" if column doesn't exist
            String cellValue = (selectedCol < row.length && !row[selectedCol].trim().isEmpty())
                               ? row[selectedCol].trim()
                               : "(empty)";

            // Increment the count for this value
            frequencyMap.put(cellValue, frequencyMap.getOrDefault(cellValue, 0) + 1);
        }

        // Sort entries by count descending for easy reading
        List<Map.Entry<String, Integer>> sortedEntries = new ArrayList<>(frequencyMap.entrySet());
        sortedEntries.sort((a, b) -> b.getValue() - a.getValue());

        int totalRows = records.size();  // Used to calculate percentages

        // ---------------------------------------------------------------
        // Step 4: Display the frequency table
        // ---------------------------------------------------------------
        System.out.println();
        System.out.println("╔══════════════════════════════════════════════════════════╗");
        System.out.println("║           MP11 – COLUMN FREQUENCY COUNTER                ║");
        System.out.println("╚══════════════════════════════════════════════════════════╝");
        System.out.println("  Source   : Pearson VUE – UPHSD Molino Exam Results");
        System.out.printf ("  Column   : %s%n", selectedColumnName);
        System.out.printf ("  Total Rows Analyzed: %d%n", totalRows);
        System.out.println("──────────────────────────────────────────────────────────");

        // Table header row
        System.out.printf("  %-54s  %6s  %7s%n", "Value", "Count", "Percent");
        System.out.println("  " + "─".repeat(70));

        for (Map.Entry<String, Integer> entry : sortedEntries) {
            String value   = entry.getKey();    // The unique value
            int count      = entry.getValue();  // How many times it appears
            double percent = (count * 100.0) / totalRows;  // Percentage of total

            System.out.printf("  %-54s  %6d  %6.2f%%%n", value, count, percent);
        }

        System.out.println("  " + "─".repeat(70));
        System.out.printf("  %-54s  %6d  %6.2f%%%n", "TOTAL", totalRows, 100.0);

        System.out.println("══════════════════════════════════════════════════════════");
        System.out.println("           End of Column Frequency Report                 ");
        System.out.println("══════════════════════════════════════════════════════════");
    }

    /**
     * parseCSVLine – Parses a single CSV line into an array of string fields.
     *
     * Handles fields enclosed in double-quotes such as "LastName,FirstName".
     * Commas inside quoted fields are treated as part of the field, not separators.
     *
     * @param line  A raw CSV line
     * @return      Array of parsed string field values
     */
    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder current = new StringBuilder();  // Builds the current field
        boolean inQuotes = false;  // Whether we're currently inside a quoted section

        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes;  // Toggle quote mode
            } else if (c == ',' && !inQuotes) {
                fields.add(current.toString());  // Field separator found; save field
                current.setLength(0);            // Reset buffer for next field
            } else {
                current.append(c);  // Regular character; add to current field
            }
        }
        fields.add(current.toString());  // Add the final field

        return fields.toArray(new String[0]);
    }
}
